src/ui/popupmenu/completion_item_widget.rs

```diff
@@ -37,11 +40,15 @@ impl CompletionItemWidgetWrap {
         let grid = gtk::Grid::new();
         grid.set_column_spacing(10);
 
-        let buf = get_icon_pixbuf(&item.kind.as_str(), icon_fg, size);
-        let kind = gtk::Image::new_from_pixbuf(&buf);
-        kind.set_tooltip_text(format!("kind: '{}'", item.kind).as_str());
-        kind.set_margin_start(margin);
-        grid.attach(&kind, 0, 0, 1, 1);
+        let image = gtk::Image::new();
+        if show_kind {
+            let buf = get_icon_pixbuf(&item.kind, icon_fg, size);
+            image.set_from_pixbuf(&buf);
+        }
+
+        image.set_tooltip_text(format!("kind: '{}'", item.kind_raw).as_str());
+        image.set_margin_start(margin);
```

Comment:
I think moving this to the `show_kind` branch above should fix the padding issue.

Reply:
@badosu It appears that the padding problem is coming directly from the LSP. Here is the output from `dbg!(&item.word)` (notice the lack of a space in the bottom completion string):

![image](https://user-images.githubusercontent.com/46855713/59713771-ca52db00-91d5-11e9-985f-ba99001c5f74.png)

So I guess we can just not worry about it then, right? (It's from the LSP? Correct me if I'm wrong.)

> Also the surrounding lines, it should not be necessary to add it to the grid or set tooltip info.

I didn't quite understand this. Do you mean if `show_kind` is false, we don't need to call those functions (`set_tooltip_text` and `grid.attach`)?

Reply:
This is for the case where all unknown items show margin. There's no reason to attach images to the popup menu when that's the case, unless if it breaks some other structure that depends on their existence.

Reply:
> @badosu It appears that the padding problem may be coming directly from the LSP (vim-lsp). Here is the output from dbg!(&item.word) (notice the lack of a space in the bottom completion string):

Also, it may not necessarily be the fault of the LSP, you need to inspect as far up on the stack as possible to confirm if we aren't changing the text by mistake (since this was probably not happening before).

Try debugging `popupmenu.items` [here](https://github.com/badosu/gnvim/blob/4b248cda4a6858544c6dbe9a4ad79796a7008247/src/ui/ui.rs#L676).

Reply:
@badosu I used `dbg!(&word)` and `dbg(&item[0])` where you said, and the output showed the same: no space for the "Set" completion string, unlike the others.

Concerning your points on attaching images only if needed, how's this diff?
```diff
diff --git a/src/ui/popupmenu/completion_item_widget.rs b/src/ui/popupmenu/completion_item_widget.rs
index 95f31b5..e77de30 100644
--- a/src/ui/popupmenu/completion_item_widget.rs
+++ b/src/ui/popupmenu/completion_item_widget.rs
@@ -44,10 +44,10 @@ impl CompletionItemWidgetWrap {
         if show_kind {
             let buf = get_icon_pixbuf(&item.kind, icon_fg, size);
             image.set_from_pixbuf(&buf);
+            image.set_tooltip_text(format!("kind: '{}'", item.kind_raw).as_str());
+            image.set_margin_start(margin);
         }
 
-        image.set_tooltip_text(format!("kind: '{}'", item.kind_raw).as_str());
-        image.set_margin_start(margin);
         grid.attach(&image, 0, 0, 1, 1);
 
         let menu = gtk::Label::new(item.menu.as_str());
```

Attaching the images to the grid either way keeps the side padding when no icons are shown.

Reply:
You forgot to move the `grid::attach` line, I just tested and it works for all unknown items:

![Screenshot from 2019-06-19 07-43-11](https://user-images.githubusercontent.com/347552/59759287-19e0e780-9266-11e9-9970-c37be1862d5d.png)

Reply:
It also works for matched items case:

![Screenshot from 2019-06-19 07-47-13](https://user-images.githubusercontent.com/347552/59759467-8825aa00-9266-11e9-9d82-80f20c12b8b9.png)

The other concern indeed seems to be an issue with the LSP.

This is the whole working diff:

```diff
diff --git a/src/ui/popupmenu/completion_item_widget.rs b/src/ui/popupmenu/completion_item_widget.rs
index 95f31b5..a21beff 100644
--- a/src/ui/popupmenu/completion_item_widget.rs
+++ b/src/ui/popupmenu/completion_item_widget.rs
@@ -44,12 +44,11 @@ impl CompletionItemWidgetWrap {
         if show_kind {
             let buf = get_icon_pixbuf(&item.kind, icon_fg, size);
             image.set_from_pixbuf(&buf);
+            image.set_tooltip_text(format!("kind: '{}'", item.kind_raw).as_str());
+            image.set_margin_start(margin);
+            grid.attach(&image, 0, 0, 1, 1);
         }

-        image.set_tooltip_text(format!("kind: '{}'", item.kind_raw).as_str());
-        image.set_margin_start(margin);
-        grid.attach(&image, 0, 0, 1, 1);
-
         let menu = gtk::Label::new(item.menu.as_str());
         menu.set_halign(gtk::Align::End);
         menu.set_hexpand(true);
```

Reply:
@badosu The reason I left the `grid.attach()` function was because of [this comment from vhakulinen]( https://github.com/vhakulinen/gnvim/pull/50#issuecomment-500590610):
> On a case when there are no icons shown (e.g. all kinds are unknown), remember to add some padding to the left side of the actual text.

Attaching the images to the grid keeps the padding on the left. Should we use less padding than when the icons are shown? Less padding may look better, but is it worth the extra logic? The normal padding amount may be too much when no icon is shown; what do you think? (See the bottom image of [this comment](https://github.com/vhakulinen/gnvim/pull/50#issuecomment-503220005) to see what it looks like with the normal padding)

Reply:
Does this looks better?

![Screenshot from 2019-06-19 14-07-38](https://user-images.githubusercontent.com/347552/59785628-a35edc80-929b-11e9-8f43-92cf292f166f.png)

I added to line 60:

```rust
if !show_kind {
  word.set_margin_start(5);
}
```

Reply:
@badosu That definitely looks much better, good idea. Should I push the change, or wait to see what vhakulinen thinks?

Reply:
By some padding, I meant exactly what @badosu demonstrated.

Reply:
Cool! I’ll push the change shortly.